### PR TITLE
Code coverage with kcov and codecov.io

### PR DESCRIPTION
**Pull request by neonphog**
looks like both `kcov` and `tarpaulin` need special docker flags to run:

```
--security-opt seccomp=unconfined
```

this makes the image download much slower and breaks the incremental build caching stuff...

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -30,6 +30,20 @@ jobs:
       command: nix-shell --run hc-rust-test
       no_output_timeout: 20m
 
+ cov-kcov:
+  machine: true
+  resource_class: xlarge
+  steps:
+   - checkout
+   - run: docker run --rm --security-opt seccomp=unconfined -v $(pwd):/project holochain/holochain-rust:circle.build.develop sh -c "cd /project && nix-shell --run hc-rust-coverage-kcov"
```
Comment (neonphog) [inline .circleci/config.yml]: both tarpaulin and kcov need this docker `--security-opt seccomp=unconfined` - I don't know another way of doing this with circleci - but this makes downloading the docker image slower.

File: codecov.yml [hunk 2]
```diff
@@ -1,4 +1,6 @@
 # Repository Yaml
+codecov:
+  token: 684bc22e-c6ca-4ac5-b5be-d1dd250a4567
```
Comment (neonphog) [inline codecov.yml]: ... committing this seems vaguely questionable, but I'm not a circle/codecov guru - don't know how else to do it.

File: rust/test/default.nix [hunk 3]
```diff
@@ -1,12 +1,70 @@
...
-  script = pkgs.writeShellScriptBin name
+  hc-rust-test = pkgs.writeShellScriptBin "hc-rust-test"
   ''
   hc-rust-wasm-compile && HC_SIMPLE_LOGGER_MUTE=1 RUST_BACKTRACE=1 cargo test --all "$1" -- --test-threads=${holonix.rust.test.threads};
   '';
+
+  hc-rust-coverage-kcov = pkgs.writeShellScriptBin "hc-rust-coverage-kcov"
+  ''
+  # we need the kcov tool and curl for fetching the codecov.io uploader
+  nix-env -f https://github.com/NixOS/nixpkgs-channels/archive/nixos-19.09.tar.gz -iA kcov curl
```
Comment (neonphog) [inline rust/test/default.nix]: This is a bit of a sledgehammer - but I keep failing at being able to put in dependencies in any other way.

File: rust/test/default.nix [hunk 4]
```diff
@@ -1,12 +1,70 @@
...
+
+  # using cargo-make for executing kcov
+  cargo install cargo-make
```
Comment (neonphog) [inline rust/test/default.nix]: again... seems like maybe this should be part of the nix environment some magical way?

File: rust/test/default.nix [hunk 5]
```diff
@@ -1,12 +1,70 @@
...
+    # remove any pre-existing target dir
+    # symlink to shared target dir
+    # build the test binaries
+    # run the code coverage
+
+    ( \
+      cd $i && \
+      rm -rf target || true && \
+      ln -s ../../target
+      export CARGO_TARGET_DIR=$(readlink -f ./target) && \
```
Comment (neonphog) [inline rust/test/default.nix]: cargo make / kcov are very particular about the placement of the target directory - this breaks incremental builds so the coverage run takes 46 minutes - but I wasn't able to get it to work in any other way.

File: rust/test/default.nix [hunk 6]
```diff
@@ -1,12 +1,70 @@
...
+
+  hc-rust-coverage-tarpaulin = pkgs.writeShellScriptBin "hc-rust-coverage-tarpaulin"
+  ''
+  unset CARGO_TARGET_DIR
+  nix-env -f https://github.com/NixOS/nixpkgs-channels/archive/nixos-19.09.tar.gz -iA curl
+  cargo install cargo-tarpaulin
+  hc-rust-wasm-compile
+  export CARGO_TARGET_DIR=$(readlink -f ./target)
+  cargo tarpaulin -v -o Xml --exclude-files "*/.cargo/*"
+  bash <(curl -s https://codecov.io/bash)
```
Comment (neonphog) [inline rust/test/default.nix]: The tarpaulin binary installs and runs just fine - but how they are doing the instrumenting does not seem to be compatible with our wasm and ffi binding - so many of our crates to not succeed in getting covered.

**Comment by neonphog** [comment]
@thedavidmeister - would love your review / help here - there are many nix-y / circle things that aren't ideal - or if we don't want to spend time on ideal this technically works and we can always refactor later.

File: .circleci/config.yml [hunk 7]
```diff
@@ -30,6 +30,20 @@ jobs:
       command: nix-shell --run hc-rust-test
       no_output_timeout: 20m
 
+ cov-kcov:
```
Comment (AshantiMutinta) [inline .circleci/config.yml]: Only comment I have is maybe we can change this to "Code Coverage" when it shows up? but definitely not necessarily at all

File: codecov.yml (hunk 2 above)
Comment (thedavidmeister) [reply codecov.yml]: added to environment

File: .circleci/config.yml (hunk 7 above)
Comment (neonphog) [reply .circleci/config.yml]: :+1: - done

File: rust/test/default.nix (hunk 4 above)
Comment (neonphog) [reply rust/test/default.nix]: we're fine sticking with this for ci task for now

**Pull request merged**